lib/elixir/test/elixir/kernel/warning_test.exs

```diff
@@ -1770,6 +1770,18 @@ defmodule Kernel.WarningTest do
     end
   end
 
+  test "defstruct warns with duplicate keys" do
+    import ExUnit.CaptureIO
+
+    assert capture_err(fn ->
+             Code.eval_string("""
+             defmodule TestMod do
+               defstruct [:foo, :foo, :bar]
+             end
+             """)
+           end) =~ "duplicate key :foo found in struct"
+  end
```

Comment:
I think it would be good to purge the module at the end of the test, just in case:

```elixir
test "defstruct warns with duplicate keys" do
  ...
after
  purge(TestMod)
end
```